### PR TITLE
[4/4] CUDNNv8 ResNet Fusion: Add fuse_resunit IR pass

### PR DESCRIPTION
**Pull request by Tom-Zheng**
### PR types

New features
### PR changes

Others
### Description

This PR adds a `fuse_resunit` IR pass, which is able to detect ResUnit patterns and replace them with CUDNNv8 fused kernels.

### CODE REVIEW

**Comment by Tom-Zheng** [comment]
@zyfncg Can we add PIR pass in the next PR and merge this first?

**Comment by Tom-Zheng** [comment]
@zyfncg CIs have passed, would you please review it?

File: paddle/fluid/framework/ir/fuse_resunit_pass.h [hunk 1]
```diff
@@ -0,0 +1,105 @@
+// Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
+// Copyright (c) 2022 NVIDIA Authors. All Rights Reserved.
```
Comment (zyfncg) [inline paddle/fluid/framework/ir/fuse_resunit_pass.h]: 2022->2023

**Review by zyfncg — DISMISSED** [review]

**Comment by zyfncg** [comment]
> @zyfncg Can we add PIR pass in the next PR and merge this first?

Ok~

**Review by zyfncg — DISMISSED** [review]

**Comment by Tom-Zheng** [comment]
```
2023-12-09 16:34:59 0. You must have raindrops2sea or XiaoguangHu01 approval for change 20+ files or add than 1000+ lines of content.
2023-12-09 16:34:59 1. Unittest is not allowed to be disabled.
2023-12-09 16:34:59 You must have one RD (kolinwei(Recommend), wanghuancoder, luotao1, QingshuChen, qili93 or ZzSean or Aurelius84) approval for the usage of @unittest.skip or @unittest.skipIf.
2023-12-09 16:34:59 +@unittest.skipIf(skip_unit_test(), skip_msg)
2023-12-09 16:34:59 +@unittest.skipIf(skip_unit_test(), skip_msg)
2023-12-09 16:34:59 +@unittest.skipIf(skip_unit_test(), skip_msg)
2023-12-09 16:34:59 +@unittest.skipIf(skip_unit_test(), skip_msg)
2023-12-09 16:34:59 +@unittest.skipIf(skip_unit_test(), skip_msg)
2023-12-09 16:34:59 +@unittest.skipIf(skip_unit_test(), skip_msg)
2023-12-09 16:34:59 +@unittest.skipIf(skip_unit_test(), skip_msg)
2023-12-09 16:34:59 2. You must have one QA (XieYunshen(Recommend) or chalsliu) approval for setting parameter RUN_TYPE as EXCLUSIVE, DIST, HYBRID, NIGHTLY, EXCLUSIVE:NIGHTLY or DISTNIGHTLY, or setting parameter SERIAL, or setting TIMEOUT properties.
2023-12-09 16:34:59 The corresponding lines are as follows:
2023-12-09 16:34:59 test/legacy_test/CMakeLists.txt
2023-12-09 16:34:59 +  set_tests_properties(test_fuse_resunit_pass PROPERTIES TIMEOUT 300)
2023-12-09 16:34:59 For more information, please refer to:https://github.com/PaddlePaddle/Paddle/wiki/PaddlePaddle-Unit-test-specification
2023-12-09 16:34:59 There are 3 approved errors.
```
@raindrops2sea @kolinwei @XieYunshen would you please approve for the above items?

**Comment by Tom-Zheng** [comment]
```
2023-12-09 16:34:59 ****************
2023-12-09 16:34:59 0. You must have raindrops2sea or XiaoguangHu01 approval for change 20+ files or add than 1000+ lines of content.
2023-12-09 16:34:59 1. Unittest is not allowed to be disabled.
2023-12-09 16:34:59 You must have one RD (kolinwei(Recommend), wanghuancoder, luotao1, QingshuChen, qili93 or ZzSean or Aurelius84) approval for the usage of @unittest.skip or @unittest.skipIf.
2023-12-09 16:34:59 +@unittest.skipIf(skip_unit_test(), skip_msg)
2023-12-09 16:34:59 +@unittest.skipIf(skip_unit_test(), skip_msg)
2023-12-09 16:34:59 +@unittest.skipIf(skip_unit_test(), skip_msg)
2023-12-09 16:34:59 +@unittest.skipIf(skip_unit_test(), skip_msg)
2023-12-09 16:34:59 +@unittest.skipIf(skip_unit_test(), skip_msg)
2023-12-09 16:34:59 +@unittest.skipIf(skip_unit_test(), skip_msg)
2023-12-09 16:34:59 +@unittest.skipIf(skip_unit_test(), skip_msg)
2023-12-09 16:34:59 2. You must have one QA (XieYunshen(Recommend) or chalsliu) approval for setting parameter RUN_TYPE as EXCLUSIVE, DIST, HYBRID, NIGHTLY, EXCLUSIVE:NIGHTLY or DISTNIGHTLY, or setting parameter SERIAL, or setting TIMEOUT properties.
2023-12-09 16:34:59 The corresponding lines are as follows:
2023-12-09 16:34:59 test/legacy_test/CMakeLists.txt
2023-12-09 16:34:59 +  set_tests_properties(test_fuse_resunit_pass PROPERTIES TIMEOUT 300)
2023-12-09 16:34:59 For more information, please refer to:https://github.com/PaddlePaddle/Paddle/wiki/PaddlePaddle-Unit-test-specification
2023-12-09 16:34:59 There are 3 approved errors.
2023-12-09 16:34:59 ****************
```
@raindrops2sea @kolinwei @XieYunshen would you please approve for above items?

**Comment by Tom-Zheng** [comment]
```
2023-12-09 17:03:28 **************************************************************
2023-12-09 17:03:28 Please find RD for approval first, and then find TPM for approval.
2023-12-09 17:03:28 0. print or std::cout is not recommended for direct use, please use logging or VLOG. If it is necessary to use, please contact tianshuo78520a (Recommend) or zhangbo9674 review and approve.
2023-12-09 17:03:28 
2023-12-09 17:03:28 There are 1 approved errors.
2023-12-09 17:03:28 **************************************************************
```
@tianshuo78520a  This may be caused by https://github.com/PaddlePaddle/Paddle/pull/59771/files#diff-5ae2e24ee75701ca9ee05b6508b45071275e22719bb35ddb8891a84cf6b8f4dcR269  
But I followed the same routine of other dist tests, like https://github.com/PaddlePaddle/Paddle/blob/70c4d213735c0408c4998a918d09fe194c9c27d4/test/distributed_passes/test_dist_fuse_gemm_epilogue_pass.py#L155 
Could you approve it?

**Review by ForFishes — DISMISSED** [review]
LGTM

**Review by XieYunshen — DISMISSED** [review]

**Comment by tianshuo78520a** [comment]
> Could you approve it?

可以先approve，下一个PR能否使用logging进行替换？

**Review by tianshuo78520a — APPROVED** [review]

**Comment by Tom-Zheng** [comment]
@raindrops2sea @XiaoguangHu01 Would you please approve for change 20+ files
@kolinwei @wanghuancoder would you please approve for usage of unittest.skip

**Comment by Tom-Zheng** [comment]
Resolved conflicts

**Comment by Tom-Zheng** [comment]
@zyfncg Would you please approve again?

**Comment by Tom-Zheng** [comment]
```
****************
2023-12-12 23:38:29 0. You must have raindrops2sea or XiaoguangHu01 approval for change 20+ files or add than 1000+ lines of content.
2023-12-12 23:38:29 1. Unittest is not allowed to be disabled.
2023-12-12 23:38:29 You must have one RD (kolinwei(Recommend), wanghuancoder, luotao1, QingshuChen, qili93 or ZzSean or Aurelius84) approval for the usage of @unittest.skip or @unittest.skipIf.
2023-12-12 23:38:29 +@unittest.skipIf(skip_unit_test(), skip_msg)
2023-12-12 23:38:29 +@unittest.skipIf(skip_unit_test(), skip_msg)
2023-12-12 23:38:29 +@unittest.skipIf(skip_unit_test(), skip_msg)
2023-12-12 23:38:29 +@unittest.skipIf(skip_unit_test(), skip_msg)
2023-12-12 23:38:29 +@unittest.skipIf(skip_unit_test(), skip_msg)
2023-12-12 23:38:29 +@unittest.skipIf(skip_unit_test(), skip_msg)
2023-12-12 23:38:29 +@unittest.skipIf(skip_unit_test(), skip_msg)
2023-12-12 23:38:29 2. You must have one QA (XieYunshen(Recommend) or chalsliu) approval for setting parameter RUN_TYPE as EXCLUSIVE, DIST, HYBRID, NIGHTLY, EXCLUSIVE:NIGHTLY or DISTNIGHTLY, or setting parameter SERIAL, or setting TIMEOUT properties.
2023-12-12 23:38:29 The corresponding lines are as follows:
2023-12-12 23:38:29 test/legacy_test/CMakeLists.txt
2023-12-12 23:38:29 +  set_tests_properties(test_fuse_resunit_pass PROPERTIES TIMEOUT 300)
2023-12-12 23:38:29 For more information, please refer to:https://github.com/PaddlePaddle/Paddle/wiki/PaddlePaddle-Unit-test-specification
2023-12-12 23:38:29 There are 3 approved errors.
2023-12-12 23:38:29 ****************
```
@XieYunshen @XiaoguangHu01 @kolinwei , @wanghuancoder would you please approve?

**Comment by Tom-Zheng** [comment]
@tianshuo78520a Would you please approve PR-CI-Coverage?

**Review by zyfncg — APPROVED** [review]

**Review by wanghuancoder — APPROVED** [review]

**Review by XieYunshen — APPROVED** [review]
LGTM

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Comment by Tom-Zheng** [comment]
@zyfncg would you please help merge it?

**Pull request merged**